From: Poul-Henning Kamp <phk@FreeBSD.org>
Date: Mon, 12 May 2014 11:29:46 +0000
Subject: Cope with systems having both <sys/endian.h> and <endian.h>

Fixes:	1476
---
 lib/libvarnish/vsha256.c | 19 +++++++++----------
 1 file changed, 9 insertions(+), 10 deletions(-)

diff --git a/lib/libvarnish/vsha256.c b/lib/libvarnish/vsha256.c
index a2208cf..515576c 100644
--- a/lib/libvarnish/vsha256.c
+++ b/lib/libvarnish/vsha256.c
@@ -28,18 +28,17 @@
 
 #include "config.h"
 
-#ifdef HAVE_SYS_ENDIAN_H
-#include <sys/types.h>
-#include <sys/endian.h>
-#define VBYTE_ORDER	_BYTE_ORDER
-#define VBIG_ENDIAN	_BIG_ENDIAN
+#if defined(HAVE_SYS_ENDIAN_H)
+#  include <sys/types.h>
+#  include <sys/endian.h>
+#  define VBYTE_ORDER	_BYTE_ORDER
+#  define VBIG_ENDIAN	_BIG_ENDIAN
+#elif defined(HAVE_ENDIAN_H)
+#  include <endian.h>
+#  define VBYTE_ORDER	__BYTE_ORDER
+#  define VBIG_ENDIAN	__BIG_ENDIAN
 #endif
 
-#ifdef HAVE_ENDIAN_H
-#include <endian.h>
-#define VBYTE_ORDER	__BYTE_ORDER
-#define VBIG_ENDIAN	__BIG_ENDIAN
-#endif
 #include <errno.h>
 #include <stdint.h>
 #include <string.h>
