Origin: WMF
Description: Fix VSM permissions
--- a/bin/varnishd/mgt/mgt_jail_unix.c
+++ b/bin/varnishd/mgt/mgt_jail_unix.c
@@ -292,15 +292,15 @@ vju_fixfd(int fd, enum jail_fixfd_e what
 
 	switch (what) {
 	case JAIL_FIXFD_FILE:
-		AZ(fchmod(fd, 0750));
+		AZ(fchmod(fd, 0755));
 		AZ(fchown(fd, vju_wrkuid, vju_wrkgid));
 		break;
 	case JAIL_FIXFD_VSMMGT:
-		AZ(fchmod(fd, 0750));
+		AZ(fchmod(fd, 0755));
 		AZ(fchown(fd, vju_uid, vju_gid));
 		break;
 	case JAIL_FIXFD_VSMWRK:
-		AZ(fchmod(fd, 0750));
+		AZ(fchmod(fd, 0755));
 		AZ(fchown(fd, vju_wrkuid, vju_wrkgid));
 		break;
 	default:
