diff --git a/lib/libvcc/vcc_expr.c b/lib/libvcc/vcc_expr.c
index 5989f07..af3b3e7 100644
--- a/lib/libvcc/vcc_expr.c
+++ b/lib/libvcc/vcc_expr.c
@@ -942,6 +942,7 @@ vcc_expr_mul(struct vcc *tl, struct expr **e, enum var_type fmt)
 	case INT:	f2 = INT; break;
 	case DURATION:	f2 = REAL; break;
 	case BYTES:	f2 = REAL; break;
+	case REAL:	f2 = REAL; break;
 	default:
 		if (tl->t->tok != '*' && tl->t->tok != '/')
 			return;
@@ -1049,6 +1050,8 @@ vcc_expr_add(struct vcc *tl, struct expr **e, enum var_type fmt)
 			/* OK */
 		} else if ((*e)->fmt == INT && e2->fmt == INT) {
 			/* OK */
+		} else if ((*e)->fmt == REAL && e2->fmt == REAL) {
+			/* OK */
 		} else if ((*e)->fmt == DURATION && e2->fmt == DURATION) {
 			/* OK */
 		} else if (tk->tok == '+' &&
